storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java

```diff
@@ -0,0 +1,494 @@
+     * @throws IOException on any error
+     * @throws InterruptedException if interrupted wile waiting for the process to exit.
+     */
+    protected boolean runProfilingCommand(List<String> command, Map<String, String> env, String logPrefix, File targetDir) throws IOException, InterruptedException {
+        Process p = Utils.launchProcess(command, env, logPrefix, null, targetDir);
+        int ret = p.waitFor();
+        return ret == 0;
+    }
+    
+    @Override
+    public boolean runProfiling(ProfileRequest request, boolean stop) throws IOException, InterruptedException {
+        if (_port <= 0) {
+            throw new IllegalStateException("Cannot profile a container recovered with just a worker id");
+        }
+        String topologyId = _assignment.get_topology_id();
+        String targetDir = ConfigUtils.workerArtifactsRoot(_conf, topologyId, _port);
+        Map<String, Object> topologyConf = ConfigUtils.readSupervisorStormConf(_conf, topologyId);
+        
+        @SuppressWarnings("unchecked")
+        Map<String, String> env = (Map<String, String>) topologyConf.get(Config.TOPOLOGY_ENVIRONMENT);
+        if (env == null) {
+            env = new HashMap<String, String>();
+        }
+
+        String str = ConfigUtils.workerArtifactsPidPath(_conf, topologyId, _port);
+
+        String workerPid = null;
+        try (FileReader reader = new FileReader(str); BufferedReader br = new BufferedReader(reader)) {
+            workerPid = br.readLine().trim();
+        }
+        
+        ProfileAction profileAction = request.get_action();
+        String logPrefix = "ProfilerAction process " + topologyId + ":" + _port + " PROFILER_ACTION: " + profileAction + " ";
+
+        List<String> command = mkProfileCommand(profileAction, stop, workerPid, targetDir);
+
+        File targetFile = new File(targetDir);
+        return runProfilingCommand(command, env, logPrefix, targetFile);
+    }
+    
+    private List<String> mkProfileCommand(ProfileAction action, boolean stop, String workerPid, String targetDir) {
+        if (action == ProfileAction.JMAP_DUMP) {
+            return jmapDumpCmd(workerPid, targetDir);
+        } else if (action == ProfileAction.JSTACK_DUMP) {
+            return jstackDumpCmd(workerPid, targetDir);
+        } else if (action == ProfileAction.JPROFILE_DUMP) {
+            return jprofileDump(workerPid, targetDir);
+        } else if (action == ProfileAction.JVM_RESTART) {
+            return jprofileJvmRestart(workerPid);
+        } else if (!stop && action == ProfileAction.JPROFILE_STOP) {
+            return jprofileStart(workerPid);
+        } else if (stop && action == ProfileAction.JPROFILE_STOP) {
+            return jprofileStop(workerPid, targetDir);
+        }
+        return Lists.newArrayList();
+    }
+
+    private List<String> jmapDumpCmd(String pid, String targetDir) {
+        return Lists.newArrayList(_profileCmd, pid, "jmap", targetDir);
```

Comment:
Not sure what you mean by this.  Ideally if we had an enum that we could put a function like cmd in.  Then It would just be

```
return action.cmd(workerPid, targetDir, stop);
```

but ProfileAction is thrift generated code so we cannot play games like that here.